ivadomed/losses.py

```diff
@@ -1,16 +1,54 @@
 import torch
 import torch.nn as nn
-import torch.nn.functional as F
+
+
+class MultiClassDiceLoss(nn.Module):
+    # Inspired from https://arxiv.org/pdf/1802.10508.pdf
+    def forward(self, prediction, target):
+        smooth = 1.0
+        dice_per_class = 0
+        n_classes = prediction.shape[1]
+        for i in range(n_classes):
+            iflat = prediction[:, i, ].reshape(-1)
+            tflat = target[:, i, ].reshape(-1)
+            intersection = (iflat * tflat).sum()
+            dice_per_class += intersection / (iflat.sum() + tflat.sum() + smooth)
+        return -(2.0 * dice_per_class) / n_classes
+
+
+class CombinedDiceLoss(nn.Module):
+    """
+    :param classes_of_interest: list containing the index of a class which dice will be added to the loss.
+    :param multiclass_dice: boolean representing whether or not multiclass dice is included in the current loss.
+    :param dice: boolean representing whether or not dice is included in the current loss.
+    """
+
+    def __init__(self, class_of_interest, multi_class_dice, dice):
+        super().__init__()
+        self.class_of_interest = class_of_interest
+        self.multi_class_dice = multi_class_dice
+        self.dice = dice
+
+    def forward(self, prediction, target):
+        loss = torch.zeros(1).cuda()
+        if len(self.class_of_interest):
+            for idx in self.class_of_interest:
+                loss += - dice_loss(prediction[:, idx, ], target[:, idx, ])
+        if self.multi_class_dice:
+            loss += MultiClassDiceLoss()(prediction, target)
+        if self.dice:
+            loss += - dice_loss(prediction, target)
+        return loss / (len(self.class_of_interest) + self.dice + self.multi_class_dice)
 
 
 def dice_loss(input, target):
     smooth = 1.0
 
-    iflat = input.view(-1)
-    tflat = target.view(-1)
+    iflat = input.reshape(-1)
+    tflat = target.reshape(-1)
```

Comment:
I was just reading: https://stackoverflow.com/questions/49643225/whats-the-difference-between-reshape-and-view-in-pytorch
Could you comment why you changed? I am not saying this is wrong, just curious :)

Reply:
view only supports contiguous data, so I had errors when using MultiClassDiceLoss because the way I used it, data was not always contiguous. I also had the error on the HeMIS branch. Since then I made changes and the data is always contiguous, but I see no disadvantage to use reshape since it supports contiguous and discontiguous data :). Should I add comment in the code as well?

Reply:
Yes sure :-)
Many thanks